File: VB.Net/Park Point Ultrasonic Sensor/Anasayfa.vb
Imports System.Security

Public Class Anasayfa
    Dim CihazAraAktif As Boolean
    Dim TarananID As String = ""
    Function Seriport_Ayarlandi() As Boolean
        TarananID = ""
        If Comport_Combo.Text = "" Then
            Return False
        End If
        If SerialPortx.IsOpen Then
            Return True
        End If


        Try

            SerialPortx.Close()
            SerialPortx.BaudRate = 38400
            SerialPortx.Parity = IO.Ports.Parity.None
            SerialPortx.StopBits = IO.Ports.StopBits.One
            SerialPortx.DataBits = 8
            SerialPortx.PortName = Comport_Combo.Text
            SerialPortx.Open()
            Portu_Ac_Buton.Text = "Kapat"
            Return True
        Catch ex As Exception
            MsgBox(ex.Message)
        End Try

        Return False
    End Function



    Private Sub Anasayfa_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Comport_Combo.Items.AddRange(System.IO.Ports.SerialPort.GetPortNames)
        If Comport_Combo.Items.Count > 0 Then
            Comport_Combo.SelectedIndex = Comport_Combo.Items.Count - 1 ' Sonuncu Olur genelde...
        End If
    End Sub

    Private Sub Yenile_Buton_Click(sender As Object, e As EventArgs) Handles Yenile_Buton.Click
        Comport_Combo.Items.Clear()
        Comport_Combo.Items.AddRange(System.IO.Ports.SerialPort.GetPortNames)
        Comport_Combo.SelectedIndex = -1
        If Comport_Combo.Items.Count > 0 Then
            Comport_Combo.SelectedIndex = Comport_Combo.Items.Count - 1 ' Sonuncu Olur genelde...
        End If
    End Sub

    Sub GelenData(ByVal Data As String)
        Try

            If Data.Length > 5 AndAlso Data.IndexOf(Chr(13)) > -1 Then

                Dim Sperator As String = Data.Substring(0, 1)
                Dim CihazID As Integer = Data.Substring(3, 3)
                Dim Komut As String = Data.Substring(1, 2)
                Cihaz_ID_Text.Value = CihazID

                If Sperator = "#" Then

                    If Komut = "DS" Then ' Device Source
                        Yeni_Cihaz_ID_Text.Value = CihazID

                        Min_Algilama_Mesafe_Text.Value = CInt(Data.Substring(6, 3))
                        Max_Algilama_Mesafe_Text.Value = CInt(Data.Substring(9, 3))
                        Kirmizi_Yakma_Text.Value = CInt(Data.Substring(12, 2))
                        Yesile_Donme_Text.Value = CInt(Data.Substring(14, 2))
                        Sensor_Durumu_Setle(CInt(Data.Substring(16, 1)))
                        Dim Versionx As String = Data.Substring(18, 5)
                        Dim Arr() As String = Versionx.Split(".")
                        Donanim_Versiyon_Label.Text = CInt(Arr(0)) & "." & CInt(Arr(1))
                        Cihaz_Ara_Buton.Text = "Cihaz Ara"
                        MsgBox("Bulunan Cihaz ID: " & CihazID)

                    ElseIf Komut = "SM" Then
                        Olculen_Mesafe_Text.Text = CInt(Data.Substring(6, 3)) & " cm"
                        Sensor_Durumu_Setle(CInt(Data.Substring(9, 1)))

                    ElseIf Komut = "SS" Then
                        Sensor_Durumu_Setle(CInt(Data.Substring(6, 1)))

                    End If





                ElseIf Sperator = "%" Then

                    If Data.IndexOf("E") > -1 Then
                        MsgBox("Hata İşlem Başarısız !", 16, "Hata")
                    Else
                        MsgBox("İşlem Başarılı !", 64, "Bilgi")
                    End If





                End If

            End If


        Catch ex As Exception
            MsgBox(ex.Message)
        End Try

    End Sub

    Private Sub Cihaz_Ara_Buton_Click(sender As Object, e As EventArgs) Handles Cihaz_Ara_Buton.Click
        If CihazAraAktif = False Then
            If Seriport_Ayarlandi() Then
                Cihaz_Ara_Timer.Start()
                CihazAraAktif = True
                Cihaz_Ara_Buton.Text = "Durdur"

            End If
        Else
            CihazAraAktif = False
            Cihaz_Ara_Timer.Stop()
            Cihaz_Ara_Buton.Text = "Cihaz Ara"
        End If



    End Sub



    Sub Sensor_Durumu_Setle(ByVal SensorDurumu As Integer)
        If SensorDurumu = 0 Then
            Sensor_Rengi_Picturebox.BackColor = Color.SpringGreen
        Else
            Sensor_Rengi_Picturebox.BackColor = Color.Tomato
        End If
    End Sub
    Dim ID As Integer
    Private Sub Cihaz_Ara_Timer_Tick(sender As Object, e As EventArgs) Handles Cihaz_Ara_Timer.Tick
        Application.DoEvents()

        If TarananID = "" Then
            ID = Cihaz_ID_Text.Value
            TarananID = ID
        End If
        Taranan_ID_Text.Text = ID.ToString

        Dim Str As String = "$DS" & ID.ToString.PadLeft(3, "0") & "**" & Chr(13)
        SerialPortx.Write(Str)

        ID += 1
        If ID >= 1000 Then
            Taranan_ID_Text.Text = "999"
            Cihaz_Ara_Timer.Stop()
            TarananID = ""
            CihazAraAktif = False
            Cihaz_Ara_Buton.Text = "Cihaz Ara"
        End If

    End Sub

    Private Sub SerialPortx_DataReceived(sender As Object, e As IO.Ports.SerialDataReceivedEventArgs) Handles SerialPortx.DataReceived
        System.Threading.Thread.Sleep(50)
        TarananID = ""
        CihazAraAktif = False
        Cihaz_Ara_Timer.Stop()
        Me.Invoke(Sub() GelenData(SerialPortx.ReadExisting))

    End Sub

    Private Sub Cihaz_ID_Degistir_Buton_Click(sender As Object, e As EventArgs) Handles Cihaz_ID_Degistir_Buton.Click

        If CihazAraAktif = False Then
            If Seriport_Ayarlandi() Then
                Dim Komut As String = "$DN"
                Dim EskiID As String = Cihaz_ID_Text.Value.ToString.PadLeft(3, "0")
                Dim YeniID As String = Yeni_Cihaz_ID_Text.Value.ToString.PadLeft(3, "0")
                Dim GonderilecekKomut As String = Komut & EskiID & YeniID & "**" & Chr(13)
                SerialPortx.Write(GonderilecekKomut)


            End If
        End If



    End Sub

    Private Sub Mesafe_Olc_Buton_Click(sender As Object, e As EventArgs) Handles Mesafe_Olc_Buton.Click
        If CihazAraAktif = False Then
            If Seriport_Ayarlandi() Then
                Dim Komut As String = "$SM"
                Dim CihazID As String = Cihaz_ID_Text.Value.ToString.PadLeft(3, "0")
                Dim GonderilecekKomut As String = Komut & CihazID & "**" & Chr(13)
                SerialPortx.Write(GonderilecekKomut)
            End If
        End If
    End Sub

    Private Sub Sensor_Algilama_Mesafe_Ayarla_Buton_Click(sender As Object, e As EventArgs) Handles Sensor_Algilama_Mesafe_Ayarla_Buton.Click
        If CihazAraAktif = False Then
            If Seriport_Ayarlandi() Then
                Dim Komut As String = "$SD"
                Dim CihazID As String = Cihaz_ID_Text.Value.ToString.PadLeft(3, "0")
                Dim MinAlgilama As String = Min_Algilama_Mesafe_Text.Value.ToString.PadLeft(3, "0")
                Dim MaxAlgilama As String = Max_Algilama_Mesafe_Text.Value.ToString.PadLeft(3, "0")
                Dim GonderilecekKomut As String = Komut & CihazID & MinAlgilama & MaxAlgilama & "**" & Chr(13)
                SerialPortx.Write(GonderilecekKomut)


            End If
        End If
    End Sub

    Private Sub Led_Durum_Degistirme_Sureleri_Buton_Click(sender As Object, e As EventArgs) Handles Led_Durum_Degistirme_Sureleri_Buton.Click
        If CihazAraAktif = False Then
            If Seriport_Ayarlandi() Then
                Dim Komut As String = "$SL"
                Dim CihazID As String = Cihaz_ID_Text.Value.ToString.PadLeft(3, "0")
                Dim KirmiziYakmaSuresi As String = Kirmizi_Yakma_Text.Value.ToString.PadLeft(2, "0")
                Dim YesileDonmeSuresi As String = Yesile_Donme_Text.Value.ToString.PadLeft(2, "0")
                Dim GonderilecekKomut As String = Komut & CihazID & KirmiziYakmaSuresi & YesileDonmeSuresi & "**" & Chr(13)
                SerialPortx.Write(GonderilecekKomut)


            End If
        End If
    End Sub

    Private Sub Led_Durum_Ogren_Buton_Click(sender As Object, e As EventArgs) Handles Led_Durum_Ogren_Buton.Click
        If CihazAraAktif = False Then
            If Seriport_Ayarlandi() Then
                Dim Komut As String = "$SS"
                Dim CihazID As String = Cihaz_ID_Text.Value.ToString.PadLeft(3, "0")
                Dim GonderilecekKomut As String = Komut & CihazID & "**" & Chr(13)
                SerialPortx.Write(GonderilecekKomut)


            End If
        End If
    End Sub

    Private Sub Portu_Ac_Buton_Click(sender As Object, e As EventArgs) Handles Portu_Ac_Buton.Click

        If Portu_Ac_Buton.Text = "Kapat" Then
            Portu_Ac_Buton.Text = "Aç"
            SerialPortx.Close()
        Else
            Seriport_Ayarlandi()
        End If

    End Sub
End Class

File: C#/Park Point Ultrasonic Sensor/Anasayfa.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Park_Point_Ultrasonic_Sensor
{
    public partial class Anasayfa : Form
    {
        private bool CihazAraAktif;
        private string TarananID = "";

        public Anasayfa()
        {
            InitializeComponent();
        }


        
        public bool Seriport_Ayarlandi()
        {
            TarananID = "";
            if (Comport_Combo.Text == "") return false;
            if (SerialPortx.IsOpen) return true;



   

            try
            {
                SerialPortx.Close();
                SerialPortx.BaudRate = 38400;
                SerialPortx.Parity = System.IO.Ports.Parity.None;
                SerialPortx.StopBits = System.IO.Ports.StopBits.One;
                SerialPortx.DataBits = 8;
                SerialPortx.PortName = Comport_Combo.Text;
                SerialPortx.Open();
                Portu_Ac_Buton.Text = "Kapat";
                return true;
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }

            return false;
        }


        private void Anasayfa_Load(object sender, EventArgs e)
        {
            Comport_Combo.Items.AddRange(System.IO.Ports.SerialPort.GetPortNames());
            if (Comport_Combo.Items.Count > 0)
                Comport_Combo.SelectedIndex = Comport_Combo.Items.Count - 1;// Sonuncu Olur genelde...
        }

        private void Yenile_Buton_Click(object sender, EventArgs e)
        {
            Comport_Combo.Items.Clear();
            Comport_Combo.Items.AddRange(System.IO.Ports.SerialPort.GetPortNames());
            Comport_Combo.SelectedIndex = -1;
            if (Comport_Combo.Items.Count > 0)
                Comport_Combo.SelectedIndex = Comport_Combo.Items.Count - 1;// Sonuncu Olur genelde...
        }


        public void GelenData(string Data)
        {
            try
            {
                if ((Data.Length > 5) && (Data.IndexOf(Convert.ToChar(13)) > -1))
                {
                    string Sperator = Data.Substring(0, 1);
                    int CihazID =Convert.ToInt32(Data.Substring(3, 3));
                    string Komut = Data.Substring(1, 2);
                    Cihaz_ID_Text.Value = CihazID;

                    if (Sperator == "#")
                    {
                        if (Komut == "DS")
                        {
                            Yeni_Cihaz_ID_Text.Value = CihazID;

                            Min_Algilama_Mesafe_Text.Value = Convert.ToInt32(Data.Substring(6, 3));
                            Max_Algilama_Mesafe_Text.Value = Convert.ToInt32(Data.Substring(9, 3));
                            Kirmizi_Yakma_Text.Value = Convert.ToInt32(Data.Substring(12, 2));
                            Yesile_Donme_Text.Value = Convert.ToInt32(Data.Substring(14, 2));
                            Sensor_Durumu_Setle(Convert.ToInt32(Data.Substring(16, 1)));
                            string Versionx = Data.Substring(18, 5);
                            string[] Arr = Versionx.Split(Convert.ToChar("."));
                            Donanim_Versiyon_Label.Text = Convert.ToInt32(Arr[0]) + "." + Convert.ToInt32(Arr[1]);
                            Cihaz_Ara_Buton.Text = "Cihaz Ara";
                            MessageBox.Show("Bulunan Cihaz ID: " + CihazID);
                        }
                        else if (Komut == "SM")
                        {
                            Olculen_Mesafe_Text.Text = Convert.ToString(Data.Substring(6, 3)) + " cm";
                            Sensor_Durumu_Setle(Convert.ToInt32(Data.Substring(9, 1)));
                        }
                        else if (Komut == "SS")
                            Sensor_Durumu_Setle(Convert.ToInt32(Data.Substring(6, 1)));
                    }
                    else if (Sperator == "%")
                    {
                        if (Data.IndexOf("E") > -1)
                            MessageBox.Show("Hata İşlem Başarısız !","Hata !",MessageBoxButtons.OK,MessageBoxIcon.Error);
                        else
                           
                        MessageBox.Show("İşlem Başarılı !", "Bilgi", MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                }
            }

            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }
        }






        public void Sensor_Durumu_Setle(int SensorDurumu)
        {
            if (SensorDurumu == 0)
                Sensor_Rengi_Picturebox.BackColor = Color.SpringGreen;
            else
                Sensor_Rengi_Picturebox.BackColor = Color.Tomato;
        }

        int ID;
        private void Cihaz_Ara_Buton_Click(object sender, EventArgs e)
        {
            if (CihazAraAktif == false)
            {
                if (Seriport_Ayarlandi())
                {
                    Cihaz_Ara_Timer.Start();
                    CihazAraAktif = true;
                    Cihaz_Ara_Buton.Text = "Durdur";
                }
            }
            else
            {
                CihazAraAktif = false;
                Cihaz_Ara_Timer.Stop();
                Cihaz_Ara_Buton.Text = "Cihaz Ara";
            }
        }

        private void Cihaz_Ara_Timer_Tick(object sender, EventArgs e)
        {
            Application.DoEvents();

            if (TarananID == "")
            {
                ID =Convert.ToInt32(Cihaz_ID_Text.Value);
                TarananID = ID.ToString();
            }
            Taranan_ID_Text.Text = ID.ToString();

            string Str = "$DS" + ID.ToString().PadLeft(3,Convert.ToChar("0")) + "**" + Convert.ToChar(13);
            SerialPortx.Write(Str);

            ID += 1;
            if (ID >= 1000)
            {
                Taranan_ID_Text.Text = "999";
                Cihaz_Ara_Timer.Stop();
                TarananID = "";
                CihazAraAktif = false;
            }
        }

        private void SerialPortx_DataReceived(object sender, System.IO.Ports.SerialDataReceivedEventArgs e)
        {
            System.Threading.Thread.Sleep(50);
            TarananID = "";
            CihazAraAktif = false;
            Cihaz_Ara_Timer.Stop();
            Invoke(new Action(() => GelenData(SerialPortx.ReadExisting())));
        }

        private void Cihaz_ID_Degistir_Buton_Click(object sender, EventArgs e)
        {
            if (CihazAraAktif == false)
            {
                if (Seriport_Ayarlandi())
                {
                    string Komut = "$DN";
                    string EskiID = Cihaz_ID_Text.Value.ToString().PadLeft(3,Convert.ToChar("0"));
                    string YeniID = Yeni_Cihaz_ID_Text.Value.ToString().PadLeft(3, Convert.ToChar("0"));
                    string GonderilecekKomut = Komut + EskiID + YeniID + "**" + Convert.ToChar(13);
                    SerialPortx.Write(GonderilecekKomut);
                }
            }
        }

        private void Mesafe_Olc_Buton_Click(object sender, EventArgs e)
        {
            if (CihazAraAktif == false)
            {
                if (Seriport_Ayarlandi())
                {
                    string Komut = "$SM";
                    string CihazID = Cihaz_ID_Text.Value.ToString().PadLeft(3, Convert.ToChar("0"));
                    string GonderilecekKomut = Komut + CihazID + "**" + Convert.ToChar(13);
                    SerialPortx.Write(GonderilecekKomut);
                }
            }
        }

        private void Sensor_Algilama_Mesafe_Ayarla_Buton_Click(object sender, EventArgs e)
        {
            if (CihazAraAktif == false)
            {
                if (Seriport_Ayarlandi())
                {
                    string Komut = "$SD";
                    string CihazID = Cihaz_ID_Text.Value.ToString().PadLeft(3, Convert.ToChar("0"));
                    string MinAlgilama = Min_Algilama_Mesafe_Text.Value.ToString().PadLeft(3, Convert.ToChar("0"));
                    string MaxAlgilama = Max_Algilama_Mesafe_Text.Value.ToString().PadLeft(3, Convert.ToChar("0"));
                    string GonderilecekKomut = Komut + CihazID + MinAlgilama + MaxAlgilama + "**" + Convert.ToChar(13);
                    SerialPortx.Write(GonderilecekKomut);
                }
            }
        }

        private void Led_Durum_Degistirme_Sureleri_Buton_Click(object sender, EventArgs e)
        {
            if (CihazAraAktif == false)
            {
                if (Seriport_Ayarlandi())
                {
                    string Komut = "$SL";
                    string CihazID = Cihaz_ID_Text.Value.ToString().PadLeft(3, Convert.ToChar("0"));
                    string KirmiziYakmaSuresi = Kirmizi_Yakma_Text.Value.ToString().PadLeft(2, Convert.ToChar("0"));
                    string YesileDonmeSuresi = Yesile_Donme_Text.Value.ToString().PadLeft(2, Convert.ToChar("0"));
                    string GonderilecekKomut = Komut + CihazID + KirmiziYakmaSuresi + YesileDonmeSuresi + "**" + Convert.ToChar(13);
                    SerialPortx.Write(GonderilecekKomut);
                }
            }
        }

        private void Led_Durum_Ogren_Buton_Click(object sender, EventArgs e)
        {
            if (CihazAraAktif == false)
            {
                if (Seriport_Ayarlandi())
                {
                    string Komut = "$SS";
                    string CihazID = Cihaz_ID_Text.Value.ToString().PadLeft(3, Convert.ToChar("0"));
                    string GonderilecekKomut = Komut + CihazID + "**" + Convert.ToChar(13);
                    SerialPortx.Write(GonderilecekKomut);
                }
            }
        }

        private void Portu_Ac_Buton_Click(object sender, EventArgs e)
        {

            if (Portu_Ac_Buton.Text == "Kapat") {

                Portu_Ac_Buton.Text = "Aç";
            SerialPortx.Close();
            }
           
           else
            {
                Seriport_Ayarlandi();
            }
            
      


        }
    }
}
